daemon/api.go

```diff
@@ -2688,3 +2701,86 @@ func postApps(c *Command, r *http.Request, user *auth.UserState) Response {
 	st.EnsureBefore(0)
 	return AsyncResponse(nil, &Meta{Change: chg.ID()})
 }
+
+type putStoreData struct {
+	Store string `json:"store"`
+}
+
+type storeResponseData struct {
+	URL string `json:"url"`
+}
+
+// Set new store API from store assertion.
+func putStore(c *Command, r *http.Request, user *auth.UserState) Response {
+	errResponse := checkRootUser(r, "configure store")
+	if errResponse != nil {
+		return errResponse
+	}
+
+	var storeData putStoreData
+	if err := json.NewDecoder(r.Body).Decode(&storeData); err != nil {
+		return BadRequest("cannot decode store data from request body: %v", err)
+	}
+	if storeData.Store == "" {
+		return BadRequest("store is required")
+	}
+
+	st := c.d.overlord.State()
+
+	st.Lock()
+	db := assertstate.DB(st)
+	st.Unlock()
+
+	assert, err := db.Find(asserts.StoreType, map[string]string{
+		"store": storeData.Store,
+	})
+	if err != nil {
+		if err == asserts.ErrNotFound {
+			return BadRequest("cannot find store assertion with store %q: %s", storeData.Store, err)
+		}
+		msg := "unexpected error finding store assertion"
+		logger.Noticef("%s: %s", msg, err)
+		return InternalError(msg)
+	}
+
+	store := assert.(*asserts.Store)
+	storeURL := store.URL()
+
+	// URL is optional, but it makes no sense to set a store with no URL.
+	if storeURL == nil {
+		return BadRequest("store assertion with store %q has no URL", storeData.Store)
+	}
+
+	// Replace active store and update state.
+	st.Lock()
+	defer st.Unlock()
+	err = storestate.SetBaseURL(st, storeURL)
```

Comment:
Should we ensure that there are no changes in flight?

Reply:
Good question! But I think we're ok - there's a single func, storestate.SetBaseURL, that replaces the current store and updates state, and that func requires a state lock to succeed.